File: Leetcode/C++/Maximal Square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        if (m==0) return 0;
        int n = matrix[0].size();

        vector<vector<int>> dp;

        //assign matrix
        for(int i=0; i<m; i++){
        	dp.push_back(vector<int> {});
        	for(int j=0; j<n;j++)
        		dp[i].push_back(matrix[i][j] - '0');
        }

        for(int i=0; i<m; i++){
        	for(int j=0; j<n; j++)
        		cout<<dp[i][j]<<' ';
        	cout<<endl;
        }

        int mx_border = 0;

        //assign dp row 0
        for(int j=0; j<n; j++)
        	mx_border = max(mx_border, dp[0][j]);

        //
        for(int i=0; i<m; i++)
        	mx_border = max(mx_border, dp[i][0]);

        for(int i=1; i<m; i++)
        	for(int j=1; j<n; j++)
        		if (dp[i][j] == 1){
        			dp[i][j] = min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) + 1;
        			mx_border = max(mx_border, dp[i][j]);
        		}

        return mx_border*mx_border;
    }
};

File: Leetcode/python/Maximal Square.py
class Solution:
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype:         """
        
        m = len(matrix)
        if (m==0): return 0
        n = len(matrix[0])
        arr = []
        dp = []

        for i in range(m):
        	dp.append([]) 
        	for j in range(n):
        		dp[i].append(ord(matrix[i][j]) - ord('0'))

        mx_border = 0 

        for j in range(n):
        	mx_border = max(mx_border, dp[0][j]) 

        for i in range(m):
        	mx_border = max(mx_border, dp[i][0]) 

        for i in range(1, m):
        	for j in range(1, n):
        		if (dp[i][j] == 1) :
        			dp[i][j] = min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) + 1 
        			mx_border = max(mx_border, dp[i][j]) 

        return mx_border*mx_border 


s = Solution()
matrix = []
matrix.append('10100')
matrix.append('10111')
matrix.append('11111')
matrix.append('10010')

print(s.maximalSquare(matrix))










